server/parsec/components/memory/invite.py

```diff
@@ -238,6 +276,104 @@ async def new_for_device(
 
             return token, send_email_outcome
 
+    @override
+    async def new_for_shamir_recovery(
+        self,
+        now: DateTime,
+        organization_id: OrganizationID,
+        author: DeviceID,
+        send_email: bool,
+        claimer_user_id: UserID,
+        # Only needed for testbed template
+        force_token: InvitationToken | None = None,
+    ) -> tuple[InvitationToken, None | SendEmailBadOutcome] | InviteNewForShamirBadOutcome:
+        try:
+            org = self._data.organizations[organization_id]
+        except KeyError:
+            return InviteNewForShamirBadOutcome.ORGANIZATION_NOT_FOUND
+        if org.is_expired:
+            return InviteNewForShamirBadOutcome.ORGANIZATION_EXPIRED
+
+        async with (
+            org.topics_lock(read=["common"]),
+            org.advisory_lock_exclusive(AdvisoryLock.InvitationCreation),
+        ):
+            try:
+                author_device = org.devices[author]
+            except KeyError:
+                return InviteNewForShamirBadOutcome.AUTHOR_NOT_FOUND
+            author_user_id = author_device.cooked.user_id
+
+            try:
+                author_user = org.users[author_user_id]
+            except KeyError:
+                return InviteNewForShamirBadOutcome.AUTHOR_NOT_FOUND
+            if author_user.is_revoked:
+                return InviteNewForShamirBadOutcome.AUTHOR_REVOKED
+
+            # Check that the claimer exists
+            claimer = org.users.get(claimer_user_id)
+            if claimer is None:
+                return InviteNewForShamirBadOutcome.USER_NOT_FOUND
+            claimer_human_handle = claimer.cooked.human_handle
+
+            # Check that a shamir setup exists
+            shamir_setup = org.shamir_setup.get(claimer_user_id)
+            if shamir_setup is None:
+                # Since the author only knows about a shamir recovery if they are part of it,
+                # we don't have a specific error for the case where the shamir setup doesn't exist
+                return InviteNewForShamirBadOutcome.AUTHOR_NOT_ALLOWED
```

Comment:
Is there a specific reason to not define a SetupNotFound error ?

Reply:
AFAIC we decided to not share the brief and removal certificates with users that are not part of the corresponding shamir. That means that users don't have a way to know whether a user has a shamir recovery setup when they are not part of it. For this reason, it would be weird to have the server give this information through an error status.